components/provider_ppucd/src/structs.rs

```diff
@@ -0,0 +1,92 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+
+use icu_uniset::UnicodeSet;
+use serde::{Deserialize, Serialize};
+use std::convert::TryInto;
+
+//
+// data key structs - the structs used directly by users of data provider
+//
+
+macro_rules! data_key {
+    (uniset, $sub_category:literal, $version:tt) => {
+        data_key!(icu_provider::DataCategory::Uniset, $sub_category, $version)
+    };
+
+    // this was copied over from `provider` crate, but maybe should be refactored
+    ($category:expr, $sub_category:literal, $version:tt) => {
+        icu_provider::DataKey {
+            category: $category,
+            sub_category: tinystr::tinystr16!($sub_category),
+            version: $version,
+        }
+    };
+}
+
+pub mod key {
+    use icu_provider::DataKey;
+    pub const WSPACE_V1: DataKey = data_key!(uniset, "WSpace", 1);
+}
+
+/// Gets a locale-invariant default struct given a data key in this module's category.
+#[cfg(feature = "invariant")]
+pub(crate) fn get_invariant(data_key: &DataKey) -> Option<DataResponse<'static>> {
+    use crate::invariant::make_inv_response;
+    match *data_key {
+        key::WSPACE_V1 => make_inv_response::<UnicodeProperty>(),
+        _ => None,
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, PartialEq, Clone)]
+pub struct UnicodeProperties {
+    pub props: Vec<UnicodeProperty>,
+}
+
+#[derive(Deserialize, Serialize, Debug, PartialEq, Clone)]
+pub struct UnicodeProperty {
+    pub name: String,
```

Comment:
Does the name need to be `String` or could it be `&'s str` (adding a lifetime parameter to the struct)?

Reply:
Done. Thanks for the help in changing owned strings to slices and lifetimes.